File: packages/node/src/sdk/events/envelope.ts
/**
 * Messaging-transport event envelope (snake_case wrapper).
 *
 * This shape matches `schemas/event-envelope.schema.json` and is intended for
 * async/messaging adapters. It is not consumed by the logger pipeline.
 */
export interface EventEnvelope {
  event_type: string;
  event_version: string;
  timestamp: string;
  payload: Record<string, unknown>;
  source?: string;
  correlation_id?: string;
  tenant_id?: string;
  metadata?: Record<string, unknown>;
}

/**
 * Fields accepted when building a messaging transport envelope.
 */
export interface EventEnvelopeFields {
  event_type: string;
  event_version?: string;
  timestamp?: string;
  payload: Record<string, unknown>;
  source?: string;
  correlation_id?: string;
  tenant_id?: string;
  metadata?: Record<string, unknown>;
}

/**
 * Build a snake_case messaging transport envelope.
 *
 * @param fields - Envelope discriminator, payload, and optional correlation metadata.
 * @returns An envelope suitable for async/messaging transports.
 */
export function buildEventEnvelope(fields: EventEnvelopeFields): EventEnvelope {
  const envelope: EventEnvelope = {
    event_type: fields.event_type,
    event_version: fields.event_version ?? '1.0.0',
    timestamp: fields.timestamp ?? new Date().toISOString(),
    payload: fields.payload,
  };

  if (fields.source !== undefined) {
    envelope.source = fields.source;
  }
  if (fields.correlation_id !== undefined) {
    envelope.correlation_id = fields.correlation_id;
  }
  if (fields.tenant_id !== undefined) {
    envelope.tenant_id = fields.tenant_id;
  }
  if (fields.metadata !== undefined) {
    envelope.metadata = fields.metadata;
  }

  return envelope;
}

File: packages/python/dt3_sdk/events/envelope.py
"""Messaging transport envelope adapter."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Dict, Mapping, Optional


def wrap_log_event(
    log_event: Mapping[str, Any],
    *,
    event_type: Optional[str] = None,
    event_version: str = "1.0.0",
    source: Optional[str] = None,
) -> Dict[str, Any]:
    """Wrap a canonical LogEvent as schemas/event-envelope.schema.json payload."""
    name = event_type or str(log_event.get("event.name", "GENERIC_EVENT"))
    return {
        "event_type": name,
        "event_version": event_version,
        "timestamp": log_event.get("timestamp")
        or datetime.now(timezone.utc).isoformat(),
        "source": source or log_event.get("service.name"),
        "correlation_id": log_event.get("correlation.id"),
        "tenant_id": log_event.get("tenant.id"),
        "payload": dict(log_event),
        "metadata": {},
    }
